Sources/SourceKitLSP/Swift/DocumentFormatting.swift

```diff
@@ -0,0 +1,236 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LanguageServerProtocol
+
+import struct TSCBasic.AbsolutePath
+import class TSCBasic.Process
+import func TSCBasic.withTemporaryFile
+
+fileprivate extension String {
+  init?(bytes: [UInt8], encoding: Encoding) {
+    let data = bytes.withUnsafeBytes { buffer in
+      guard let baseAddress = buffer.baseAddress else {
+        return Data()
+      }
+      return Data(bytes: baseAddress, count: buffer.count)
+    }
+    self.init(data: data, encoding: encoding)
+  }
+}
+
+/// If a parent directory of `fileURI` contains a `.swift-format` file, return the path to that file.
+/// Otherwise, return `nil`.
+private func swiftFormatFile(for fileURI: DocumentURI) -> AbsolutePath? {
+  guard var path = try? AbsolutePath(validating: fileURI.pseudoPath) else {
+    return nil
+  }
+  repeat {
+    path = path.parentDirectory
+    let configFile = path.appending(component: ".swift-format")
+    if FileManager.default.isReadableFile(atPath: configFile.pathString) {
+      return configFile
+    }
+  } while !path.isRoot
+  return nil
+}
+
+/// Return the contents of a swift-format configuration file that reflects `options`.
+private func swiftFormatConfigurationFile(for options: FormattingOptions) throws -> String {
+  // The following options are not supported by swift-format and ignored:
+  // - trimTrailingWhitespace: swift-format always trims trailing whitespace
+  // - insertFinalNewline: swift-format always inserts a final newline to the file
+  // - trimFinalNewlines: swift-format always trims final newlines
+
+  if options.insertSpaces {
+    return """
+      {
+        "version": 1,
+        "tabWidth": \(options.tabSize),
+        "indentation": { "spaces": \(options.tabSize) }
+      }
+      """
+  } else {
+    return """
+      {
+        "version": 1,
+        "tabWidth": \(options.tabSize),
+        "indentation": { "tabs": 1 }
+      }
+      """
+  }
+}
+
+/// If a `.swift-format` file is discovered that applies to `fileURI`, execute `body` with the URL of that file.
+///
+/// Otherwise, create a temporary file with configuration parameters from `options` and invoke `body` with the URL of
+/// that file. The temporary file will automatically be deleted after `body` finishes.
+private func withSwiftFormatConfiguration<T>(
+  for fileURI: DocumentURI,
+  options: FormattingOptions,
+  body: (AbsolutePath) async throws -> T
+) async throws -> T {
+  if let configFile = swiftFormatFile(for: fileURI) {
+    // If we find a .swift-format file, we ignore the options passed to us by the editor.
+    // Most likely, the editor inferred them from the current document and thus the options
+    // passed by the editor are most likely less correct than those in .swift-format.
+    return try await body(configFile)
+  } else {
```

Comment:
I'd prefer no `else`. Does `swift-format` allow setting up options without a file so there would be no need for the temporary file?

Reply:
We added the ability to pass a raw JSON string in the `--configuration` flag a while back: https://github.com/apple/swift-format/pull/634

Reply:
Oh, nice. I didn’t know that. That makes the implementation a lot nicer. 😍